KEP-3299: Update kmsv2 kep for beta (#3804)
kubernetes/enhancements: keps/sig-auth/3299-kms-v2-improvements/README.md

+    APIVersion string `json:"apiVersion"`
```

> Support key hierarchy in KMS plugin that generates local KEK and add v2alpha1 `KeyManagementService` proto service contract in Kubernetes to include `key_id`, `annotations`, and `status`.

Reviewer: do we need to mention v2beta1 here?

Author: updated it to just `v2`

##### Prerequisite testing updates

##### Unit tests

- Validate key hierarchy behavior in reference implementation
  - Local KEK is generated for encryption
  - Local KEK is rotated after a period of time/number of operations
  - Remote KEK is only called when
    - Cache is empty, so remote KEK is needed to decrypt local KEK
    - Local KEK is generated or rotated
  - When key hierarchy is not used, remote KEK is called for every encryption/decryption
  - Ensure the logs and metrics are generated as expected
  - At least 75% code coverage
- Staleness check based on keyID in the `StatusResponse`
- Unit test for gRPC request/response validation
  - Serialize/Deserialize `EncryptedObject`
  - Validate keyID in `StatusResponse` and `EncryptResponse`
  - Validate annotations in `EncryptResponse`
  - Validate logs are generated with the correct `UID` and additional metadata

##### Integration tests

- Integration tests to validate
  - Encryption of custom resources and custom resource definitions

>   - No-op writes cause rewrite of stale data (data that has correct schema but was encrypted with not the latest keyID)

Reviewer: ```suggestion
  - No-op writes cause rewrite of stale data (data that has correct schema but was encrypted with keyID that is not the latest)
```

  - Health checks
    - single health check for v2 at `/kms-providers`
    - individual health checks for v1 and v2 with `/kms-provider-0` and `/kms-provider-1`
- Integration tests with base64 plugin to validate the encryption and decryption of data
- Integration tests with reference implementation
  - Validate the encryption and decryption of data
  - Validate the functionality of reference implementation
- Integration tests to check rotation is possible without restarting API server
- Integration tests that exercise the feature enablement/disablement flow

##### e2e tests

With this e2e test suite, we want to do the following:
1. Run the e2e suite against a kind cluster without kms encryption enabled.
2. Run the e2e suite against a kind cluster that has kms v2 encryption enabled (as defined below).
3. Compare `request_duration_seconds`, `request_terminations_total`, `request_aborts_total` API server metrics between the two runs. The acceptable delta should be less than 20%.
4. Observe metrics from the reference implementation to determine time taken at each step of the encryption/decryption process.

> 5. Observe API server startup time with and without kms encryption enabled.

Author: Since we defined the acceptable delta for API server metrics (latency) to be 20% we can do the same here. WDYT @enj?

> 5. Observe API server startup time with and without kms encryption enabled.

Author: @logicalhan Me and @enj had a conversation and this is the rough math we came up with:

Assuming 10ms for every KMS RPC request to be complete, for every 5000 resource encrypted, it will increase the startup time by roughly a minute (5000*10ms = 50s). Is this something you would like us to documented in the KEP?

}
```

**Beta**
```go
FeatureSpec{
	Default: true,
	LockToDefault: false,
	PreRelease: featuregate.Beta,
}
```

###### Does enabling the feature change any default behavior?

No. The v2 API is new in the v1.25 release. Furthermore, even with the feature enabled by default, the user needs to explicitly configure a KMSv2 provider to use this.

###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?

Yes, To disable encryption at rest using the v2 API:
1. Add new `identity` provider at the top of encryption config
1. Restart kube-apiserver

> 1. Run storage migration to migrate all the existing encrypted data to use the `identity` provider

Author: > Ah, does adding identity provider unencrypt the data in etcd?

By setting `identity` as the first provider in the list, that'll be used for writes (no encryption). The rest of the providers will be used for read. When the user runs `kubectl get secrets --all-namespaces -o json | kubectl replace -f -`, the kms providers will be used to decrypt the data and on write the `identity` provider will be used. At the end of this, all the data will be stored unencrypted. 

> How do you know if the migration has successfully completed?

We're proposing a metric to record the number of times a key is used for read and write (xref: https://github.com/kubernetes/kubernetes/pull/115394). With this metric, when the API server is restarted after storage migration, the count for key used to read should be 0 indicating all the data is already unencrypted.

> 1. Run storage migration to migrate all the existing encrypted data to use the `identity` provider

Reviewer 2: Have you considered using a condition on the CRD to denote that the unencryption has completed?

> 1. Run storage migration to migrate all the existing encrypted data to use the `identity` provider

Author: We don't have a CRD for KMS. As a first step, we were thinking about exposing enough details in the metrics + documentation on how to use the metric, that could be used to determine if the rotation is complete.

> 1. Run storage migration to migrate all the existing encrypted data to use the `identity` provider

Reviewer 3: we still have the storage-version-migrator.  explaining how to verify the writes with that API is worthwhile.

1. Remove the KMS provider from the encryption config and restart kube-apiserver
2. At the end of these steps, all the data in etcd will be unencrypted.

More details are available [here](https://kubernetes.io/docs/tasks/administer-cluster/kms-provider/#disabling-encryption-at-rest)

> Disabling this gate without first doing a storage migration to use a different encryption at rest mechanism will result in data loss.

Author: > It's not just data loss, right? It will basically break the cluster, since apiserver won't be able to read the objects from etcd.

That is only when all the resources in etcd are encrypted using KMS. Typically the configuration is only for a subset of resources (secrets, configmaps). In that scenario, those resources can't be retrieved if the KMS providers are removed without running storage migration. If the user adds the KMS providers back (only for read), they would still be able to retrieve the resources.

> Disabling this gate without first doing a storage migration to use a different encryption at rest mechanism will result in data loss.

Reviewer 3: Let's imagine we encrypted secrets that are mounted in pods. This will prevent any future pods referencing secrets from starting, right?  Given that secrets are the most likely thing to encrypt, it's worth calling out the consequences of the failure here.

}
```

**Beta**
```go
FeatureSpec{
	Default: true,
	LockToDefault: false,
	PreRelease: featuregate.Beta,
}
```

###### Does enabling the feature change any default behavior?

No. The v2 API is new in the v1.25 release. Furthermore, even with the feature enabled by default, the user needs to explicitly configure a KMSv2 provider to use this.

###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?

Yes, To disable encryption at rest using the v2 API:
1. Add new `identity` provider at the top of encryption config
1. Restart kube-apiserver
1. Run storage migration to migrate all the existing encrypted data to use the `identity` provider

>    1. If running `kubectl get <resource> --all-namespaces -o json | kubectl replace -f -` to migrate data, the user can confirm that the migration is complete by observing the kube-apiserver metric `key_id_hash_last_timestamp_seconds` which will no longer contain the keyID hash of the old KEK.

Reviewer: nit:
```suggestion
   1. If running `kubectl get <resource> --all-namespaces -o json | kubectl replace -f -` to migrate data, the user can confirm that the migration is complete by observing the kube-apiserver metric `key_id_hash_last_timestamp_seconds` which will no longer have updated timestamps for the keyID hash of the old KEK.
```

I believe metrics with the old keyID hash value might still be there but the timestamp of that metric should no longer be updated.

Author: > I believe metrics with the old keyID hash value might still be there but the timestamp of that metric should no longer be updated.

I missed mentioning post API server restart here (i'll update it). Post API server restart, there should be no metric for the old keyID in timestamp or counter as it's no longer used.

>    1. If running `kubectl get <resource> --all-namespaces -o json | kubectl replace -f -` to migrate data, the user can confirm that the migration is complete by observing the kube-apiserver metrics `key_id_hash_last_timestamp_seconds` and `apiserver_envelope_encryption_key_id_hash_total`. These metrics will no longer contain the keyID hash of the old KEK after storage migration and kube-apiserver restart.

Reviewer: ```suggestion
   1. If running `kubectl get <resource> --all-namespaces -o json | kubectl replace -f -` to migrate data, the user can confirm that the migration is complete by observing the kube-apiserver metrics `apiserver_envelope_encryption_key_id_hash_last_timestamp_seconds` and `apiserver_envelope_encryption_key_id_hash_total`. These metrics will no longer contain the keyID hash of the old KEK after storage migration and kube-apiserver restart.
```

### Rollout, Upgrade and Rollback Planning

###### How can a rollout or rollback fail? Can it impact already running workloads?

- If a rollback of the feature is done without first doing a storage migration to use a different encryption at rest mechanism will result in data loss.
  - Workloads relying on existing data in etcd will no longer be able to access it.
  - The data can be retrieved by reenabling the feature gate or deleting and recreating the data.
- The rollout of the feature can fail if there are too many calls to the external kms provider.
  - API server will not report healthy.
- For highly-available clusters, the feature can be enabled on some API servers only for read purpose.
  - For rollout, add KMSv2 providers as read across all API servers first before adding the provider for write.
  - For rollback, move KMSv2 providers from write to read position across all API servers.

###### What specific metrics should inform a rollback?

- Latency metrics `transformation_duration_seconds`

> - Transformation error count metric `apiserver_storage_transformation_duration_seconds_bucket{status="Error"}`

Reviewer: ```suggestion
- Transformation error count metric `apiserver_storage_transformation_duration_seconds_bucket{transformation_type="from_storage",transformer_prefix="k8s:enc:kms:v2:}"
```

`Status` was removed: https://github.com/kubernetes/kubernetes/pull/115394#discussion_r1094339742

> - After rollback is complete, you should no longer see the keyID metric `apiserver_envelope_encryption_keyid_total` increment.

Reviewer: ```suggestion
- After rollback is complete, you should no longer see the keyID metric `apiserver_envelope_encryption_key_id_hash_total` increment.
```

- [x] Other (treat as last resort)
  - Details:
    - Logs in kube-apiserver, kms-plugin and KMS will be logged with the corresponding `key_id`, `annotations`, and `UID`.
    - Number of times a keyID is used for encryption/decryption

>     - Metric recording the current keyID that's being used as returned in the `StatusResponse`

Reviewer: ```suggestion
    - Metric recording the last time in seconds when a keyID was returned in the `StatusResponse` e.g. `apiserver_envelope_encryption_key_id_hash_status_last_timestamp_seconds{key_id_hash="sha256", provider_name="providerName"} 1.674865558833728e+09`
```
